matscipy/cauchy_born.py

```diff
@@ -0,0 +1,742 @@
+        Parameters
+        ----------
+        atoms : ASE atoms object
+            Atoms object where atoms sit at the coordinates needed to generate the deformation gradient tensor field
+            from F_func.
+        F_func : Function
+            Function to calculate the deformation gradient tensor field from a set of coordinates. The system of coordinates
+            that the function accepts is given by 'coordinates' and the function also accepts anything extra specified in *args and
+            **kwargs. Function must return the deformation gradient field for the atoms in form of a numpy array shape [natoms,ndims,ndims],
+            where F is specified in cartesian coordinates.
+        coordinates: string
+            The coordinates system that F_func accepts. Must be 'cylind2D', 'cylind3D', 'spherical', 'cart2D', 'cart3D'
+        cell : array_like, optional
+            An optional argument allowing the user to specify the cell parameters used to define the coordinate system.
+
+        Returns
+        -------
+        F3D : array_like
+            The 3D deformation gradient tensor field for all the atoms in the system, expressed in the lab frame.
+        '''
+        natoms = len(atoms)
+        if cell is None:
+            sx, sy, sz = atoms.cell.diagonal()
+        else:
+            [sx,sy,sz] = cell
+
+        if coordinates=='cylind2D':
+            x, y = atoms.positions[:, 0], atoms.positions[:, 1]
+            cx, cy = sx/2, sy/2
+            r = np.sqrt((x - cx)**2 + (y - cy)**2)
+            theta = np.arctan2((y-cy),(x-cx))
+            F_2D = F_func(r,theta,*args,**kwargs)
+            #pad F into a 3x3 matrix
+            F_3D = (np.zeros([natoms,3,3]))
+            F_3D[:,0:2,0:2] = F_2D[:,:,:]
+            F_3D[:,2,2] = 1
+
+        elif coordinates=='cylind3D':
+            x, y, z = atoms.positions[:,0],atoms.positions[:,1],atoms.positions[:,2]
+            cx, cy, cz = sx/2, sy/2, sz/2
+            r = np.sqrt((x - cx)**2 + (y - cy)**2)
+            theta = np.arctan2((y-cy),(x-cx))
+            F_3D = F_func(r,theta,z,*args,**kwargs)
+
+        elif coordinates=='spherical':
+            raise NotImplementedError
+
+        elif coordinates=='cart2D':
+            x, y = atoms.positions[:, 0], atoms.positions[:, 1]
+            F_2D = F_func(x,y,*args,**kwargs)            
+            F_3D = (np.zeros([natoms,3,3]))
+            F_3D[:,0:2,0:2] = F_2D[:,:,:]
+            F_3D[:,2,2] = 1
+        
+        elif coordinates=='cart3D':
+            x, y, z = atoms.positions[:, 0], atoms.positions[:, 1], atoms.positions[:,2]
+            F_3D = F_func(x,y,z,*args,**kwargs)
+            
+        else:
+            print('Coordinate system should be cart, cylind2D, cylind3D, spherical')
+            raise NotImplementedError
+
+        return F_3D
+
+    
+    def evaluate_F_or_E(self,A,atoms,F_func=None,E_func=None,cell=None,coordinates='cart3D',*args,**kwargs):
```

Comment:
There is not much common code between the F and E cases. Maybe better to split into two functions, once which evaluates F and one that evaluates E?